index.js

```diff
@@ -147,43 +166,55 @@ module.exports = async function (url, options) {
 
     return new Promise(function (resolve, reject) {
         const wrappedFetch = async () => {
-            ++attempt;
-
-            let timeoutHandler;
-            if (retryOptions.socketTimeout) {
-                const controller = new AbortController();
-                timeoutHandler = setTimeout(() => controller.abort(), retryOptions.socketTimeout);
-                options.signal = controller.signal;
-            }
-
-            try {
-                const response = await fetch(url, options);
-                clearTimeout(timeoutHandler);
-
-                if (!retry(retryOptions, null, response)) {
-                    // response.timeout should reflect the actual timeout
-                    response.timeout = retryOptions.socketTimeout;
-                    return resolve(response);
-                }
-
-                console.error(`Retrying in ${retryOptions.retryInitialDelay} milliseconds, attempt ${attempt - 1} failed (status ${response.status}): ${response.statusText}`);
-            } catch (error) {
-                clearTimeout(timeoutHandler);
-
-                if (!retry(retryOptions, error, null)) {
-                    if (error.name === 'AbortError') {
-                        return reject(new FetchError(`network timeout at ${url}`, 'request-timeout'));
+            let processLastResult = ()=>{};
+            while (!isResponseTimedOut(retryOptions)) {
+                ++attempt;
+
+                let timeoutHandler;
+                if (retryOptions.socketTimeout) {
+                    const controller = new AbortController();
+                    timeoutHandler = setTimeout(() => controller.abort(), retryOptions.socketTimeout);
+                    options.signal = controller.signal;
+                }                
+    
+                try {
+                    // console.log(`Fetching ${url} attempt ${attempt}`);
+                    const response = await fetch(url, options);
+                    processLastResult = ()=>resolve(response);
+
+                    if (shouldRetry(retryOptions, null, response)) {
+                        console.error(`Retrying in ${retryOptions.retryInitialDelay} milliseconds, attempt ${attempt} failed (status ${response.status}): ${response.statusText}`);
+                    } else {
+                        // response.timeout should reflect the actual timeout
+                        response.timeout = retryOptions.socketTimeout;
+                        return resolve(response);
                     }
-
-                    return reject(error);
+                } catch (error) {
+                    processLastResult = ()=>reject(error);
```

Comment:
to keep compatibility, I think we still need to pass through the special error when its an `AbortError`, wdyt?

Reply:
I think AbortError is generated by our own timeout that we're imposing -- so I'm not 100% sure about that.

Reply:
I'm not sure there is any special handling of ABortError in our code, but since this library is open source and may be used in other projects... At least, to be safe, if we keep it like this let's make a new major version.